iroh-resolver/src/chunker/rabin.rs

```diff
@@ -0,0 +1,916 @@
+        // two parts: Part A contains the result of the modulus operation, part
+        // B is used to cancel out the 8 top bits so that one XOR operation is
+        // enough to reduce modulo Polynomial
+        for b in 0u64..256 {
+            mod_table[b as usize] = modulus(b << DEG_TARGET, pol) | b << DEG_TARGET;
+        }
+
+        (out_table, mod_table)
+    }
+
+    /// The degree of the polynomial.
+    const fn deg(pol: u64) -> i64 {
+        64 - (pol.leading_zeros() as i64) - 1
+    }
+
+    /// Modulus result of numerator%denominator.
+    /// see https://en.wikipedia.org/wiki/Division_algorithm
+    fn modulus(mut numerator: u64, denominator: u64) -> u64 {
+        assert_ne!(denominator, 0, "division by zero");
+        if numerator == 0 {
+            return 0;
+        }
+
+        let denom_deg = deg(denominator);
+
+        let mut deg_diff = deg(numerator) - denom_deg;
+        while deg_diff >= 0 {
+            numerator ^= denominator << usize::try_from(deg_diff).unwrap();
+            deg_diff = deg(numerator) - denom_deg;
+        }
+
+        numerator
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use futures::TryStreamExt;
+    use proptest::prelude::*;
+    use rand::{rngs::StdRng, SeedableRng};
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_simple_chunks() {
+        let chunker = Rabin::new(Config::default(), GO_IPFS_V0_PRESET);
+
+        println!("1 chunk");
+        let data: Vec<u8> = (0..1024u32).flat_map(|i| i.to_le_bytes()).collect();
+        let bytes = std::io::Cursor::new(&data);
+        let split = chunker.clone().chunks(bytes);
+        let result: Vec<_> = split.try_collect().await.unwrap();
+        assert_eq!(result.len(), 1);
+        assert_eq!(result[0], &data);
+
+        println!(">= 4 chunks");
+        let data: Vec<u8> = (0..393_216u32).flat_map(|i| i.to_le_bytes()).collect();
+        test_rabin_roundtrip_data(data).await;
+    }
+
+    #[tokio::test]
+    async fn test_more_chunks() {
+        let seed = 1;
+        let mut rng = StdRng::seed_from_u64(seed);
+        for size in [
+            0,
+            100,
+            1024,
+            1024 * 512,
+            1024 * 1024,
+            1024 * 1024 * 5,
+            1024 * 1024 * 100,
+        ] {
+            println!("size {size}");
+            let mut data = vec![0u8; size];
+            rng.fill_bytes(&mut data);
+            test_rabin_roundtrip_data(data).await;
+            println!("----");
+        }
+    }
+
+    async fn test_rabin_roundtrip_data(data: Vec<u8>) {
+        let config = Config::default();
+        let chunker = Rabin::new(config.clone(), GO_IPFS_V0_PRESET);
+
+        let mut rng = StdRng::seed_from_u64(0);
+        // split into randomized chunks
+        let mut chunks: Vec<Vec<u8>> = Vec::new();
+        let mut index = 0;
+        while index < data.len() {
+            let split = rng.gen_range(index..=data.len());
```

Comment:
This will create smaller and smaller chunks, instead of uniformly distributed chunks. Is that intentional?

Reply:
not really, but it is fine